### PR TITLE
distsql: ignore misplanned ranges when planning locally

### PR DESCRIPTION
**Pull request by jordanlewis**
cc @nvanbenschoten 

Release note: None

### CODE REVIEW

File: pkg/sql/distsqlrun/processors.proto [hunk 1]
```diff
@@ -227,6 +227,11 @@ message TableReaderSpec {
   // consumer of this TableReader expects to be able to see in-progress schema
   // changes.
   optional ScanVisibility visibility = 7 [(gogoproto.nullable) = false];
+
+  // If true, instructs the TableReader to not check for misplanned table
+  // readers. Should be set if the TableReader is planned on a node known not
+  // to be the range's leaseholder.
+  optional bool ignore_misplanned_ranges = 8 [(gogoproto.nullable) = false];
```
Comment (asubiotto) [inline pkg/sql/distsqlrun/processors.proto]: Don't think we should be making this part of the spec. This is only relevant for local `tableReader`s. Not sure if there is another way to pass it down to a processor. Maybe a more general flag on the `flowCtx`?

**Review by asubiotto — CHANGES_REQUESTED** [review]

File: pkg/sql/distsqlrun/processors.proto (hunk 1 above)
Comment (jordanlewis) [reply pkg/sql/distsqlrun/processors.proto]: Isn't it also relevant for tableReaders that we know we're not planning correctly, like in the case of the softLimit planning or for unhealthy nodes? I didn't add those to this PR, but I feel like there could be other uses for this. If not, I agree with you and I'll revert.

**Review by asubiotto — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsql_physical_planner.go, line 1092 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWWf2s-TZUyVhiJv2M:-LNWWf2s-TZUyVhiJv2N:b-p3gii1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsql_physical_planner.go#L1092)):*
> ```Go
> 		*tr = spec
> 		tr.Spans = makeTableReaderSpans(sp.Spans)
> 		tr.IgnoreMisplannedRanges = ignoreMisplannedRanges
> ```

I would set this directly to `planCtx.isLocal`
___
*[pkg/sql/distsqlrun/processors.proto, line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWUOzUWv1nFXa0RnN7-r1-234:-LNWXC_J5kUlSZLFTSZd:b-hu5y39) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsqlrun/processors.proto#L234)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Isn't it also relevant for tableReaders that we know we're not planning correctly, like in the case of the softLimit planning or for unhealthy nodes? I didn't add those to this PR, but I feel like there could be other uses for this. If not, I agree with you and I'll revert.
</blockquote></details>

Another alternative is to have the distsqlreceiver not call `updateCaches` when it receives metadata with `Ranges`.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsql_physical_planner.go, line 1092 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWWf2s-TZUyVhiJv2M:-LNWY5cg8OvHh_ud8pWm:b-nv4uth) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsql_physical_planner.go#L1092)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would set this directly to `planCtx.isLocal`
</blockquote></details>

See my other comment - what about the other cases where we want to ignore misplanned rangeS?
___
*[pkg/sql/distsqlrun/processors.proto, line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWUOzUWv1nFXa0RnN7-r1-234:-LNWY9UQEFG4kHbnBSH7:b-844wmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsqlrun/processors.proto#L234)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Another alternative is to have the distsqlreceiver not call `updateCaches` when it receives metadata with `Ranges`.
</blockquote></details>

Yeah, but that way the remotes still have to generate the ranges. I'd rather have them avoid doing that work at all.

**Review by asubiotto — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsql_physical_planner.go, line 1092 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWWf2s-TZUyVhiJv2M:-LNW_txnBlTGi0rl1Gik:b-947bfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsql_physical_planner.go#L1092)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

See my other comment - what about the other cases where we want to ignore misplanned rangeS?
</blockquote></details>

Sure, but at this stage we're only ignoring misplanned ranges when `planCtx.isLocal`, right? I would only add the boolean once we actually need it for more complicated logic.
___
*[pkg/sql/distsqlrun/processors.proto, line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWUOzUWv1nFXa0RnN7-r1-234:-LNWZi8f91-srPkpATRz:b-gofwut) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsqlrun/processors.proto#L234)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yeah, but that way the remotes still have to generate the ranges. I'd rather have them avoid doing that work at all.
</blockquote></details>

I think misplanned ranges are kept track of regardless by the kv fetcher. You might avoid appending to a slice and possibly sending one producer metadata message but I'm not sure it's worth it to add more stuff to the spec. The distsql receiver stuff is less invasive and can be changed down the line to this approach if we want to. Up to you.

**Review by jordanlewis — COMMENTED** [review]
PTAL, I just decided to remove the bool from the spec and only pass it in local flows. I think that's fine for now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-:-LNWzqwy2Ifx_1fgx6AE:b-65dqpa)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/processors.proto, line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-LNWUOzUWv1nFXa0RnN7-r1-234:-LNWbBh197qP3GQTqh6w:b-2zg0bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c98093f2ce012b7e4c679c7c38ddf117eebe7e3/pkg/sql/distsqlrun/processors.proto#L234)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think misplanned ranges are kept track of regardless by the kv fetcher. You might avoid appending to a slice and possibly sending one producer metadata message but I'm not sure it's worth it to add more stuff to the spec. The distsql receiver stuff is less invasive and can be changed down the line to this approach if we want to. Up to you.
</blockquote></details>

I'll use the planCtx

**Review by jordanlewis — COMMENTED** [review]
This will need a backport

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-:-LNX0iSW5WFtB9SQfGF7:bikn8jk)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by asubiotto — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30775#-:-LNX3HlN9Pbncwi-VLu-:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by RaduBerinde** [comment]
CC @andreimatei 

I think you want to also pass `returnRangeInfo=false` to `fetcher.Init`. We are still requesting the data from KV, we're just not using it.

**Pull request merged**